Poison Message Handler

tanaysingh3484 opened the pull request:
- Handles the situation when a Requestor puts a message into the queue, following which the backend container goes down, and then it comes back up, resulting in a poison message.
- The logic which I have applied is that when the poison message is consumed, we perform a `MQOPEN` on the `replyToQ` , and if a `MQRC_UNKNOWN_OBJECT` is raised, we know that the reply to queue does not exist anymore, and we can conclude that the message is indeed a poison message.
- I implemented a Promise which resolves to true if the reply to queue exists, else resolves to false. This is then handled in the `Responder` file, where a false value is dealt with a simple output message, and a `resolve(null)` value.

tanaysingh3484 commented:
Currently in the `Responder` , in this block : 
```
else if(messages[0].replyToMsg) {
                    this.mqclient.checkQueueExists(messages[0].replyToMsg, this.appId)
                    .then((flag) => {
                        if (flag === false) {
                            debug_info("Some Poison Message");
                            resolve(messages);
                        } else {
                            debug_info(`Responder ${this.myID} obtained message from queue`);
                            resolve(messages);
                        }
                    })
                    .catch((err) => {
                        debug_warn("Some error has occured : ", err);
                        reject(err);
                    })
                }
```
Currently if flag is false, we are only `resolve(null)`, which does show a poison message, but the message still remains inside the queue. We could add a `resolve(messages)` , which would remove the message from the queue.

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,28 @@ class Responder {
             .then((messages) => {
                 debug_warn(`The producer retrieved this message ${JSON.stringify(messages)}`);
-                if(!messages[0].replyToMsg) {
+                if(messages.length === 0) {
```
please use house style.

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,28 @@ class Responder {
+                if(messages.length === 0) {
+                    debug_info("Poison Message");
+                    resolve();
+                } else if(!messages[0].replyToMsg) {
```
house style

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,28 @@ class Responder {
+                } else if(!messages[0].replyToMsg) {
                     debug_warn("This is not a reply to queue valid message");
                     resolve(null);
+                } else if(messages[0].replyToMsg) {
```
house style

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/msms/message-session-manager.js:
```diff
@@ -90,6 +90,50 @@ class MQClient {
+              od.ObjectType = MQC.MQOT_Q;
+              let openOptions = MQC.MQOO_OUTPUT;
+
+              mq.Open(hConn, od, openOptions, function (err, hObj) {
```
Is there a connection leak here? The queue is being opened here, but not closed. It will be opened again when the reply to message is to be written. For every message there is a connection leak.

tanaysingh3484 commented:
- Updated the `checkQueueExists` function to not have a connection leak by calling an `mq.Close` .
- Updated the copyright dates for both the updated files.
- Moved the `resolve(messages)` to be at the end of the if-else block.

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,27 @@ class Responder {
-                if(!messages[0].replyToMsg) {
+                if (messages.length === 0) {
+                    debug_info("Poison Message");
+                    resolve();
```
line 39 is resolving empty, 42 is resolving null. Do they have a different meaning?

Shouldn't 42 be resolving so the message is removed from the queue.

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,27 @@ class Responder {
+                        } else {
+                            debug_info(`Responder ${this.myID} obtained message from queue`);
+                        }
+                        resolve(messages);
```
If we are resolving the same way regardless of whether the reply to queue exists, what is the impact of invoking `checkQueueExists` ?

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,30 @@ class Responder {
+                } else if (messages[0].replyToMsg) {
+                    this.mqclient.checkQueueExists(messages[0].replyToMsg, this.appId)
+                    .then((flag) => {
+                        if (flag === false) {
```
this is a boolean so can use `if (!flag) {`

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/models/Responder.js:
```diff
@@ -34,12 +34,30 @@ class Responder {
+                } else if (!messages[0].replyToMsg) {
                     debug_warn("This is not a reply to queue valid message");
                     resolve(null);
+                } else if (messages[0].replyToMsg) {
```
This is an array of messages, does the check need to be applied to each message in the array?

tanaysingh3484 replied:
So I tried out a scenario where a Poison Message is in the queue, and a new fresh Request is sent into the queue. The responder only consumes one message from the queue each time the component is switched "on". So every time a message is consumed, its an array with a single element inside. So that's why we only check for `messages[0]`

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/msms/message-session-manager.js:
```diff
@@ -90,6 +90,58 @@ class MQClient {
+ * @param {Object} appId The current session id.
+ * @returns {Promise<any>} Returns a promise resolving to true if the reply to queue exists, else resolves with a false.
+ */
+  checkQueueExists(replyToMsg, appId) {
```
This method checks if a queue exists, it is not specific to a reply queue. I suggest changing the name, and comments, to make it less specific.

chughts commented on ibm-messaging-mq-cloud-showcase-app/backend/AWS-showcase-backend/msms/message-session-manager.js:
```diff
@@ -90,6 +90,57 @@ class MQClient {
+ * @param {Object} replyToMsg The reply to queue value present in the message body. (This is present in the message structure of the Put function)
+ * @returns {Promise<any>} Returns a promise resolving to true if the reply to queue exists, else resolves with a false.
+ */
+  checkReplyToQueueExists(replyToMsg) {
```
My request was to make this method name less specific. IE. `CheckQueueExisits`
